src/strbuf.c

```diff
@@ -0,0 +1,483 @@
+    }
+}
+
+void Strbuf_CopyChars(Strbuf *dst, const Charcode *src)
+{
+    Strbuf_Check(dst);
+
+    dst->size = 0;
+
+    while (*src != EOS) {
+        if (dst->size >= (dst->maxsize - 1)) {
+            GF_ASSERT(0);
+            break;
+        }
+
+        dst->data[dst->size++] = *src++;
+    }
+
+    dst->data[dst->size] = EOS;
+}
+
+void Strbuf_CopyNumChars(Strbuf *dst, const Charcode *src, u32 num)
+{
+    Strbuf_Check(dst);
+
+    if (num <= dst->maxsize) {
+        memcpy(dst->data, src, num * sizeof(Charcode));
+
+        u32 i;
+        for (i = 0; i < num; i++) {
+            if (dst->data[i] == EOS) {
+                break;
+            }
+        }
+
+        dst->size = i;
+        if (i == num) {
+            dst->data[num - 1] = EOS;
+        }
+
+        return;
+    }
+
+    GF_ASSERT(0);
+}
+
+void Strbuf_ToChars(const Strbuf *src, Charcode *dst, u32 dstsize)
+{
+    Strbuf_Check(src);
+
+    if ((src->size + 1) <= dstsize) {
+        memcpy(dst, src->data, (src->size + 1) * sizeof(Charcode));
+        return;
+    }
+
+    GF_ASSERT(0);
+}
+
+// Strbuf_GetData
+const Charcode* Strbuf_GetData(const Strbuf *src)
+{
+    Strbuf_Check(src);
+
+    return src->data;
+}
+
+void Strbuf_Concat(Strbuf *dst, const Strbuf *src)
+{
+    Strbuf_Check(dst);
+    Strbuf_Check(src);
+
+    if ((dst->size + src->size + 1) <= dst->maxsize) {
+        memcpy(dst->data + dst->size, src->data, (src->size + 1) * sizeof(Charcode));
+        dst->size += src->size;
+        return;
+    }
+
+    GF_ASSERT(0);
+}
+
+// Strbuf_Append
+void Strbuf_Append(Strbuf *dst, Charcode c)
+{
+    Strbuf_Check(dst);
+
+    if ((dst->size + 1) < dst->maxsize) {
+        dst->data[dst->size++] = c;
+        dst->data[dst->size] = EOS;
+        return;
+    }
+
+    GF_ASSERT(0);
+}
+
+#define COMPRESSED_LEADER       (0xF100)
+#define COMPRESSED_MASK         (0x01FF)
+#define COMPRESSED_EOS          (0x01FF) // 0xFFFF & 0x01FF
+
+// Strbuf_IsTrainerName
+BOOL Strbuf_IsTrainerName(Strbuf *strbuf)
+{
+    return strbuf->size > 0 && strbuf->data[0] == COMPRESSED_LEADER;
+}
+
+void Strbuf_ConcatTrainerName(Strbuf *dst, Strbuf *src)
+{
+    if (Strbuf_IsTrainerName(src)) {
+        Charcode *dstC = &dst->data[dst->size];
+        Charcode *srcC = &src->data[1];
+        s32 bit = 0;
+        s32 outsize = 0;
+        Charcode cur = 0;
+
+        while (TRUE) {
+            cur = (*srcC >> bit) & COMPRESSED_MASK;
+            bit += 9;
+
+            if (bit >= 15) {
+                srcC++;
+                bit -= 15;
+
+                if (bit) {
+                    cur |= (*srcC << (9 - bit)) & COMPRESSED_MASK;
+                }
+            }
+
+            if (cur == COMPRESSED_EOS) {
+                break;
+            }
+
+            *dstC++ = cur;
+            outsize++;
+        }
+
+        *dstC = EOS;
+        dst->size += outsize;
+    } else {
+        Strbuf_Concat(dst, src);
+    }
+}
+
+void Strbuf_UpperChar(Strbuf *strbuf, int i)
```

Comment:
`i` probably fine? maybe `index`, not sure.

Reply:
Think `i` is fine here. It's a generically-understood index var.